### DISCUSSION TITLE
Groupid for the OSGi Technology project

### DISCUSSION THREAD

**Discussion by maho7791** [osgi/osgi#596] [General]
I would organize OSSRH access as well as GPG setup for Github for the OSGi TEchnology project.

I had the groupid **org.eclipse.osgitech** and sub groups in mind. 

Eclipse has these conventions:
https://www.eclipse.org/projects/handbook/#resources-maven
https://wiki.eclipse.org/IT_Infrastructure_Doc#Publish_to_Maven_Central

It feels not appropriate to me publishing implementations under der **org.osgi** groupid, but a discussion on slack showed, that this could also be an option.
In this case we have to clarify the current access rules.

I want to discuss this, because it should be clarified for future contributions as well.

**Comment by laeubi** [comment]
> I had the groupid org.eclipse.osgitech and sub groups in mind.

I would choose `org.eclipse.osgi` as the base group ID and use `org.eclipse.osgi.technology`, this would follow the eclipserules and for an implementation one can the choose `org.eclipse.osgi.technology.rest` for example.

Please note that any diverging from the eclipse rules will require PMC approval!

**Reply by timothyjward** [reply]
> I would choose `org.eclipse.osgi` as the base group

As I’ve mentioned previously, this coordinate is used by Equinox (as an artifact id) and so using it risks significant confusion. 

I think that `org.eclipse.osgitech` (as suggested by @maho7791) is a better option for the base.

I agree that individual projects should simply append to the base.

**Reply by laeubi** [reply]
What is the "risk" beside that Equinox should not use this namespace according to the eclipse project rules?
I already have opened:
- https://github.com/eclipse-equinox/equinox/discussions/283

choosing a cripled group name (!) like "osgitech", is a really bad choice and should not be made just to account for other history choices.

In general the **artifact id** within a group can be freely choosen, so I dont see any issue with that another artifact is named the same in a different group-id, that's why we have "namespaces" there.

**Reply by maho7791** [reply]
Regarding the Eclipse project rules. Our project is:
https://projects.eclipse.org/projects/technology.osgi-technology

So, if I understand the rules correct, the groupid would be: **org.eclipse.osgi-technology**.

The name osgitech would then just be a shorter form.

I would like to summarize the result from this discussion in the next spec call

**Comment by maho7791** [comment]
As information: Creating OSS credentials and gpg stuff with corresponding setup for e.g. Github can be requested via Eclipse issue, as long as the groupid is **org.eclipse.<something>**

In case of **org.osgi** we would need to clarify the current state.

**Comment by merks** [comment]
Someone has been publishing here:

https://repo1.maven.org/maven2/org/osgi/
https://repo1.maven.org/maven2/org/eclipse/osgi/

Also this:

https://repo1.maven.org/maven2/org/eclipse/org.eclipse.eclipse/

What a mess.

Other people/projects even published into my project's namespace without ever asking:

![image](https://github.com/osgi/osgi/assets/208716/328de6ee-19a6-45b0-a992-b62e6f588dd9)

**Reply by maho7791** [reply]
The stuff that has been deployed under groupId **org.eclipse.osgi** seems to come from the Equinox project. They now obviously use the **org.eclipse.platform** groupId

The groupId org.osgi usually contains stuff related to the OSGi specifications, like API's or specified implemenations. This will stay the same. 

Usually implementations for specifications can be hold whereever they like. There are 

The OSGi Technology project is part of the OSGi Workinggroup and was meant as technical project that can contain code beside the pure specification part. I see this project as a possibility to hold default implementations for specifiaction, that have no homebase, like e.g. Apache Aries or a specific Eclipse project like Equinox or ECF. This project also make it possible for us to release artifacts outside the specification process. That makes sense, because implementations have their own lifecycle and need to be bugfixed.  In the current case we talk about an implementation of theJakarta RESTful Web Services Whiteboard specification. We hope that other contributions will follow, why need a groupId.

I personally do not want do disadvantage other implementations just by using the namespace **org.osgi**, which can make the impression, that this is the best implementation, which might be not the case. From that point we decided to take more neutral project  names for such kind of implementations of specification. Just because everyone is allowed to implement an OSGi specification.

After discussions about that, we decided the package and project naming **org.eclipse.osgitech**, whereas OSGi API's will be still under the **org.osgi** namespace. 

But to deploy our implementation to Maven Central, we would then also need a neutral groupId like **org.eclipse.osgitech**.
Finally, I would like to avoid any mess with competing groupIds you mentioned.

**Reply by laeubi** [reply]
There is actually nothing to avoid as you mentiond eclipse (equinox) is not longer using that group id anyways, if you want to be sure I already have suggested `org.eclipse.osgi.technology` what is unused as of today. osgitech simply sounds odd to me especially if the rationale is purely because the name was used accidentally many years ago...

**Reply by maho7791** [reply]
Got it, anyway I would leave this discussion open, at least until the next spec call. So we can gather other impressions. 
Nothing is rejected ;-)
The final decisions come from the Eclipse Admins, who have to create the setup for our Gibhub repo ;-)

**Comment by merks** [comment]
Another thing you likely want to consider is what Maven coordinates are people using to specify dependencies on OSGi's Maven artifacts today?

Looking here for example:

https://repo1.maven.org/maven2/org/eclipse/platform/org.eclipse.equinox.event/1.6.200/org.eclipse.equinox.event-1.6.200.pom

we see this:

> 
> <dependency>
> <groupId>org.osgi</groupId>
> <artifactId>org.osgi.service.event</artifactId>
> <version>1.4.1</version>
> <optional>false</optional>
> </dependency>

It seems a lot of people would need to find and use a new location and that different groups might mix and match such dependencies.  Wouldn't that be bad?

**Reply by laeubi** [reply]
`org.osgi` is mostly for the OSGi _Specification_, while this (maybe `org.eclipse.osgi.technology`) is for _implementations_ of specifications.

**Comment by merks** [comment]
@maho7791 

Thank you for the explanation!  That makes perfect sense indeed!!

**Comment by juergen-albert** [comment]
Waynes answer from quite some time ago:

What would be least confusing for consumers and adopters?

Just in case you're assuming otherwise, the EF has no issue with the OSGi projects continuing to use the "org.osgi" namespace.

I think that "org.eclipse.osgi.*" is used by the Eclipse Equinox, which may further complicate things.

The "org.eclipse.<project>.<component>" is more of a convention than a hard rule. The convention is to use the project's "short name", which in the case of OSGi Technology is "osgi-technology", meaning that following this convention would yield the package "org.eclipse.osgi-technology.*"., which feels a bit wordy, albeit pretty unambiguous.

Keeping with the convention of "org.osgi.<component>" seems like the least confusing option to me.

Is any of this helpful? The PMC may have some wisdom to offer.

Wayne

**Comment by laeubi** [comment]
Can't we use "org.eclipse.technology.osgi" ??
@juergen-albert suggested "org.osgi.technology" ... maybe even "org.osgi.technology.implementation"

**Comment by maho7791** [comment]
We started a poll with 4 options:

https://github.com/osgi/osgi/discussions/599
